### DISCUSSION TITLE
about edge update, before message passing for updating nodes

### DISCUSSION THREAD

**Discussion by choiwanuk** [pyg-team/pytorch_geometric#2132] [Q&A] [answered]
Hi, I'm trying to write codes for interaction-network in [Relational inductive biases, deep learning, and graph networks] paper.
what that I confused in pytorch-geometric on process is below image process.
![image](https://user-images.githubusercontent.com/54091448/108294555-66b65780-71d9-11eb-92fd-52ed80206364.png)

How can I update with edge with MLP like message passing and update process of nodes in PyG?

I didn't found about edge operation in documents of PyG.
Thank you for reading.

**Comment by wsad1** [comment, marked as answer]
Looks like an edge update depends on 3 things: embedding of nodes it connects, the edges previous embedding and the global graph attribute.
Specifically for this operation implementing a custom MessagePassing class might not be the most straightforward thing, as you'd have to modify `message`, `aggregate` and `update`. You could implement this in the forward function.
```
def forward(v,edge_index,e,u):
    '''
    v:node embedding.
    e: edge embedding.
    u: graph embedding.
    '''

    v_i=v.index_select(dim=0,index=edge_index[1])#get target node embedding
    v_j=v.index_select(dim=0,index=edge_index[0])#get source node embedding
    e_new=any_function(e,v_i,v_j,u)#coumpute new edge embedding. 
    
    #after this do message passing on nodes if needed.
```

**Reply by choiwanuk** [reply]
thank you for giving advices! It's very helpful  :)
